File: backend/java/src/main/java/ai/snippetquiz/core_service/shared/domain/bus/event/DomainEvent.java
package ai.snippetquiz.core_service.shared.domain.bus.event;

import java.util.HashMap;
import java.util.UUID;

public abstract class DomainEvent extends BaseEvent {
    public DomainEvent(UUID aggregateId, UUID userId) {
        super(aggregateId, userId);
    }

    public DomainEvent(
            UUID aggregateId,
            UUID userId,
            UUID eventId,
            String occurredOn,
            Integer version) {
        super(aggregateId, userId, eventId, occurredOn, version);
    }

    protected DomainEvent() {
        super();
    }

    public DomainEvent fromPrimitives(
            UUID aggregateId,
            UUID userId,
            HashMap<String, Object> body,
            UUID eventId,
            String occurredOn,
            Integer version) {
        throw new UnsupportedOperationException("Not supported yet.");
    }
}

File: backend/nestjs/apps/commons/event-bus/domain-event.ts
import { v4 as uuidv4 } from 'uuid';

export abstract class DomainEvent {
    readonly aggregateId: string;
    readonly eventId: string;
    readonly occurredOn: string;
    readonly eventName: string;
    readonly userId: string;
    readonly meta: Record<string, any>;

    constructor(
        eventName: string,
        aggregateId: string,
        userId: string,
        eventId?: string,
        occurredOn?: string,
        meta: Record<string, any> = {},
    ) {
        this.aggregateId = aggregateId;
        this.eventId = eventId || uuidv4();
        this.occurredOn = occurredOn || new Date().toISOString();
        this.eventName = eventName;
        this.userId = userId;
        this.meta = meta;
    }

    abstract toPrimitives(): DomainEventAttributes;

    static fromPrimitives(
        body: DomainEventAttributes,
        eventId: string,
        occurredOn: string,
    ): DomainEvent {
        throw new Error('Method not implemented.');
    }
}

export type DomainEventAttributes = {
    aggregate_id: string;
    user_id: string;
    [key: string]: any;
};

export interface DomainEventEnvelope {
    data: {
        event_id: string;
        version?: number;
        type: string;
        occurred_on: string;
        attributes: DomainEventAttributes;
    };
    meta: Record<string, any>;
}
